libs/injection/bbhnet/injection/injection.py

```diff
@@ -0,0 +1,261 @@
+
+
+def get_snr(data, noise_psd, fs, fmin=20):
+    """ Calculate the waveform SNR given the background noise PSD"""
+
+    data_fd = np.fft.rfft(data) / fs
+    data_freq = np.fft.rfftfreq(len(data)) * fs
+    dfreq = data_freq[1] - data_freq[0]
+
+    noise_psd_interp = noise_psd.interpolate(dfreq)
+    noise_psd_interp[noise_psd_interp == 0] = 1.0
+
+    snr = 4 * np.abs(data_fd) ** 2 / noise_psd_interp.value * dfreq
+    snr = np.sum(snr[fmin <= data_freq])
+    snr = np.sqrt(snr)
+
+    return snr
+
+
+def generate_gw(
+    sample_params,
+    ifo,
+    waveform_generator=None,
+    get_snr=False,
+    noise_psd=None,
+    whiten_fn=None,
+):
+    """Generate gravitational-wave events
+
+    Arguments:
+    - sample_params: dictionary of GW parameters
+    - sample_duration: time duration of each sample
+    - triggers: trigger time (relative to `sample_duration`) of each sample
+    - ifo: interferometer
+    - waveform_generator: bilby.gw.WaveformGenerator with appropriate params
+    - get_snr: return the SNR of each sample
+    - noise_psd: background noise PSD used to calculate SNR the sample
+    - whiten_fn: whiten each sample using the background noise PSD
+    """
+
+    n_sample = len(sample_params["geocent_time"])
+
+    if waveform_generator is None:
+        waveform_generator = bilby.gw.WaveformGenerator(
+            duration=8,
+            sampling_frequency=16384,
+            frequency_domain_source_model=lal_binary_black_hole,
+            parameter_conversion=convert_to_lal_binary_black_hole_parameters,
+            waveform_arguments={
+                "waveform_approximant": "IMRPhenomPv2",
+                "reference_frequency": 50,
+                "minimum_frequency": 20,
+            },
+        )
+
+    sample_rate = waveform_generator.sampling_frequency
+    waveform_duration = waveform_generator.duration
+    waveform_size = int(sample_rate * waveform_duration)
+
+    signals = np.zeros((n_sample, waveform_size))
+    snr = np.zeros(n_sample)
+
+    ifo = bilby.gw.detector.get_empty_interferometer(ifo)
+    for i in range(n_sample):
+        # Get parameter for one signal
+        p = dict()
+        for k, v in sample_params.items():
+            p[k] = v[i]
+        ra, dec, geocent_time, psi = (
+            p["ra"],
+            p["dec"],
+            p["geocent_time"],
+            p["psi"],
+        )
+
+        # Generate signal in IFO
+        polarizations = waveform_generator.time_domain_strain(p)
+        signal = np.zeros(waveform_size)
+        b, a = sig.butter(
+            N=8,
+            Wn=waveform_generator.waveform_arguments["minimum_frequency"],
+            btype="highpass",
+            fs=waveform_generator.sampling_frequency,
+        )
+        for mode in polarizations.keys():
+            # Get ifo response
+            response = ifo.antenna_response(ra, dec, geocent_time, psi, mode)
+            polarizations[mode] = sig.filtfilt(b, a, polarizations[mode])
```

Comment:
This can be done up front too, right? Then you could just have this loop look like:
```python
for mode, polarization in polarizations.items():
    response = ifo.antenna_response(ra, dec, geocent_time, psi, mode)
    signal += polarization * response
```

Or am I missing something and this actually gets set differently each time?

Reply:
Sorry this comment is meant to refer specifically to the line `polarizations[mode] = sig.filtfilt(b, a, polarizations[mode])`, which as I'm typing I'm realizing that these polarizations are being generated inside each loop so this can't be done up front. Still, I don't think resetting the key is necessary within the loop, and it might be cleaner as
```python
for mode, polarization in polarizations.items():
    response = ifo.antenna_response(ra, dec, geocent_time, psi, mode)
    signal += response * sig.filtfilt(b, a, polarization)

Reply:
This is cleaner, and I've made this change